app/src/main/java/com/example/habitsmasher/UserValidator.java

```diff
@@ -0,0 +1,153 @@
+package com.example.habitsmasher;
+
+import android.util.Patterns;
+import android.widget.EditText;
+
+import androidx.core.util.PatternsCompat;
+
+public class UserValidator {
+    private static final String USERNAME_IS_REQUIRED_MESSAGE = "Username is required!";
+    private static final String USERNAME_LENGTH_MUST_BE_LESS_THAN_16_MESSAGE = "Username length must be less than 16!";
+    private static final String EMAIL_IS_REQUIRED_MESSAGE = "Email is required!";
+    private static final String PASSWORD_IS_REQUIRED_MESSAGE = "Password is required!";
+    private static final String INVALID_EMAIL_FORMAT_MESSAGE = "Invalid email format!";
+    private static final String PASSWORD_LENGTH_MUST_BE_GREATER_THAN_6_MESSAGE = "Password length must be greater than 6!";
+    private static final String PASSWORD_LENGTH_MUST_BE_LESS_THAN_16_MESSAGE = "Password length must be less than 16!";
+
+    private EditText _usernameInput;
+    private EditText _emailInput;
+    private EditText _passwordInput;
+
+    String _username;
+    String _email;
+    String _password;
+
+    // for testing
+    public UserValidator() {}
+
+    // for sign up
+    public UserValidator(EditText usernameInput,
+                         EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = usernameInput;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = _usernameInput.getText().toString().trim();
+    }
+
+    // for log in
+    public UserValidator(EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = null;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = "";  // since there is no username for logging in, empty
+    }
+
+    public String getValidUsernameForSignUp() {
+        if (isNewUserValid()) {
+            return _username;
+        }
+        return "";
+    }
+    public String getValidEmailForSignUp() {
+        if (isNewUserValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForSignUp() {
+        if (isNewUserValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    public String getValidEmailForLogin() {
+        if (isLoginValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForLogin() {
+        if (isLoginValid()) {
+            return _password;
+        }
+        return "";
+    }
```

Comment:
What exactly are the purpose of these methods? They seem to be used only once within the code (not including unit tests) and they only are used to pass a valid field into a method.

Reply:
I see. Though my concern is that these methods are only called in sections when it is certain that a user/login is valid,  so an additional check when retrieving the strings is redundant.

Reply:
I see what you're saying, and it holds true for **this** stage of the project. But, in the future, if someone wants to go in and use the validator **without** properly validating beforehand, fetching these values from a "validator" implies that the data is valid.

I don't think it hurts so much to check once again to be sure before returning the value to the caller.